// @athenna/artisan/src/testing/plugins/command/TestCommand.js
/**
 * @athenna/artisan
 *
 * (c) João Lenon <lenon@athenna.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { Artisan } from '#src';
import { Assert } from '@japa/assert';
import { Path } from '@athenna/common';
import { TestOutput } from '#src/testing/plugins/command/TestOutput';
export class TestCommand {
    constructor() {
        /**
         * Japa assert class instance.
         */
        this.assert = new Assert();
    }
    /**
     * The Artisan file path that will be used to run commands.
     *
     * @default Path.bootstrap(`console.${Path.ext()}`)
     */
    static { this.artisanPath = Path.bootstrap(`console.${Path.ext()}`); }
    /**
     * Set the artisan file path.
     */
    static setArtisanPath(path) {
        this.artisanPath = path;
        return this;
    }
    /**
     * Instantiate TestOutput class from stdout, stderr and exitCode.
     */
    createOutput(output) {
        return new TestOutput(this.assert, output);
    }
    /**
     * Run the command and return the TestOutput instance
     * to make assertions.
     */
    async run(command, options) {
        return Artisan.callInChild(command, {
            path: TestCommand.artisanPath,
            ...options
        }).then(output => this.createOutput(output));
    }
}
