// @adonisjs/drive/build/standalone.js
"use strict";
/*
 * @adonisjs/drive
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathPrefixer = exports.DirectoryListing = exports.LocalFileServer = exports.FakeDriver = exports.LocalDriver = exports.DriveManager = exports.FakeDrive = void 0;
__exportStar(require("./src/Exceptions"), exports);
var Fake_1 = require("./src/Fake");
Object.defineProperty(exports, "FakeDrive", { enumerable: true, get: function () { return Fake_1.FakeDrive; } });
var DriveManager_1 = require("./src/DriveManager");
Object.defineProperty(exports, "DriveManager", { enumerable: true, get: function () { return DriveManager_1.DriveManager; } });
var Local_1 = require("./src/Drivers/Local");
Object.defineProperty(exports, "LocalDriver", { enumerable: true, get: function () { return Local_1.LocalDriver; } });
var Fake_2 = require("./src/Drivers/Fake");
Object.defineProperty(exports, "FakeDriver", { enumerable: true, get: function () { return Fake_2.FakeDriver; } });
var LocalFileServer_1 = require("./src/LocalFileServer");
Object.defineProperty(exports, "LocalFileServer", { enumerable: true, get: function () { return LocalFileServer_1.LocalFileServer; } });
var DirectoryListing_1 = require("./src/DirectoryListing");
Object.defineProperty(exports, "DirectoryListing", { enumerable: true, get: function () { return DirectoryListing_1.DirectoryListing; } });
var PathPrefixer_1 = require("./src/PathPrefixer");
Object.defineProperty(exports, "PathPrefixer", { enumerable: true, get: function () { return PathPrefixer_1.PathPrefixer; } });
