// @adonisjs/profiler/build/src/DummyProfiler/index.js
"use strict";
/*
 * @adonisjs/profiler
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.dummyRow = exports.dummyAction = void 0;
/**
 * Dummy action is a noop implementation of [[ProfileActionContract]]. When
 * actions for certain labels are disabled, then dummy action is used and
 * helps in avoiding the need of unneccessary `if/else` clauses.
 */
class DummyAction {
    end() { }
}
/**
 * Dummy row is a noop implementation of [[ProfilerRowContract]]. When certain
 * labels are disabled, then dummy row is used and helps in avoiding the need
 * of unneccessary `if/else` clauses.
 */
class DummyRow {
    constructor() {
        this.action = new DummyAction();
    }
    get hasParent() {
        return false;
    }
    profile(_action, _data, cb) {
        if (typeof cb === 'function') {
            return cb();
        }
        return this.action;
    }
    async profileAsync(_action, _data, cb) {
        if (typeof cb === 'function') {
            return cb();
        }
        return this.action;
    }
    create() {
        return this;
    }
    end() { }
}
const dummyAction = new DummyAction();
exports.dummyAction = dummyAction;
const dummyRow = new DummyRow();
exports.dummyRow = dummyRow;
