// @athenna/core/src/Commands/MakeExceptionCommand.js
/**
 * @athenna/core
 *
 * (c) João Lenon <lenon@athenna.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Path } from '@athenna/common';
import { sep, resolve, isAbsolute } from 'node:path';
import { BaseCommand, Argument } from '@athenna/artisan';
export class MakeExceptionCommand extends BaseCommand {
    static signature() {
        return 'make:exception';
    }
    static description() {
        return 'Make a new exception file.';
    }
    async handle() {
        this.logger.simple('({bold,green} [ MAKING EXCEPTION ])\n');
        const file = await this.generator
            .path(this.getFilePath())
            .template('exception')
            .setNameProperties(true)
            .make();
        this.logger.success(`Exception ({yellow} "${file.name}") successfully created.`);
    }
    /**
     * Get the file path where it will be generated.
     */
    getFilePath() {
        return this.getDestinationPath().concat(`${sep}${this.name}.${Path.ext()}`);
    }
    /**
     * Get the destination path for the file that will be generated.
     */
    getDestinationPath() {
        let destination = Config.get('rc.commands.make:exception.destination', Path.exceptions());
        if (!isAbsolute(destination)) {
            destination = resolve(Path.pwd(), destination);
        }
        return destination;
    }
}
__decorate([
    Argument({
        description: 'The exception name.',
    })
], MakeExceptionCommand.prototype, "name", void 0);
