// @athenna/core/src/Commands/MakeTestCommand.js
/**
 * @athenna/core
 *
 * (c) João Lenon <lenon@athenna.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Path } from '@athenna/common';
import { sep, resolve, isAbsolute } from 'node:path';
import { BaseCommand, Argument, Option } from '@athenna/artisan';
export class MakeTestCommand extends BaseCommand {
    static signature() {
        return 'make:test';
    }
    static description() {
        return 'Make a new test file.';
    }
    async handle() {
        this.logger.simple('({bold,green} [ MAKING TEST ])\n');
        let template = 'test';
        if (this.isFunction) {
            template = 'testFn';
        }
        const file = await this.generator
            .path(this.getFilePath())
            .template(template)
            .setNameProperties(true)
            .make();
        this.logger.success(`Test ({yellow} "${file.name}") successfully created.`);
    }
    /**
     * Get the file path where it will be generated.
     */
    getFilePath() {
        return this.getDestinationPath().concat(`${sep}${this.name}.${Path.ext()}`);
    }
    /**
     * Get the destination path for the file that will be generated.
     */
    getDestinationPath() {
        let destination = Config.get('rc.commands.make:test.destination', this.isUnit ? Path.tests('unit') : Path.tests('e2e'));
        if (!isAbsolute(destination)) {
            destination = resolve(Path.pwd(), destination);
        }
        return destination;
    }
}
__decorate([
    Argument({
        description: 'The test name.',
    })
], MakeTestCommand.prototype, "name", void 0);
__decorate([
    Option({
        description: 'Create an unitary test.',
        signature: '-u, --unit',
        default: false,
    })
], MakeTestCommand.prototype, "isUnit", void 0);
__decorate([
    Option({
        description: 'Create the test as function instead of class.',
        signature: '--function',
        default: false,
    })
], MakeTestCommand.prototype, "isFunction", void 0);
