// @athenna/artisan/src/commands/ConfigureCommand.js
/**
 * @athenna/artisan
 *
 * (c) João Lenon <lenon@athenna.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { extname, resolve } from 'node:path';
import { Argument, BaseCommand } from '#src';
import { Path, File, Module } from '@athenna/common';
import { NotFoundLibraryException } from '#src/exceptions/NotFoundLibraryException';
import { NotFoundConfigurerException } from '#src/exceptions/NotFoundConfigurerException';
export class ConfigureCommand extends BaseCommand {
    static signature() {
        return 'configure';
    }
    static description() {
        return 'Configure one or more libraries inside your application.';
    }
    async handle() {
        for (const library of this.libraries) {
            await this.configure(library);
        }
    }
    async configure(library) {
        this.logger.simple(`({bold,green} [ CONFIGURING ${library} ])\n`);
        const isFile = this.isFile(library);
        const isInstalled = await this.isInstalled(library);
        if (!isFile && !isInstalled) {
            throw new NotFoundLibraryException(library);
        }
        const path = isFile
            ? resolve(library)
            : Path.nodeModules(`${library}/configurer/index.js`);
        if (!(await File.exists(path))) {
            throw new NotFoundConfigurerException(path, library);
        }
        const Configurer = await Module.getFrom(path);
        await new Configurer().setPath(path).configure();
    }
    isFile(library) {
        return !!extname(library);
    }
    async isInstalled(library) {
        const json = await new File(Path.pwd('package.json')).getContentAsJson();
        if (json.dependencies[library]) {
            return true;
        }
        return !!json.devDependencies[library];
    }
}
__decorate([
    Argument({
        signature: 'libraries...',
        description: 'One or more libraries to be configured. (Example: node artisan configure @athenna/mail @athenna/database)'
    }),
    __metadata("design:type", Array)
], ConfigureCommand.prototype, "libraries", void 0);
