// @athenna/http/src/Context/Request.js
/**
 * @athenna/http
 *
 * (c) João Lenon <lenon@athenna.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { Is, Json } from '@athenna/common';
export class Request {
    constructor(request) {
        this.request = request;
    }
    /**
     * Get the request id.
     *
     * @example 12345
     */
    get id() {
        return this.request.id;
    }
    /**
     * Get the request ip.
     *
     * @example 192.168.0.1
     */
    get ip() {
        return this.request.ip;
    }
    /**
     * Get the request hostname.
     *
     * @example localhost
     */
    get hostname() {
        return this.request.hostname;
    }
    /**
     * Get the request protocol.
     *
     * @example http
     */
    get protocol() {
        return this.request.protocol;
    }
    /**
     * Get the request method.
     *
     * @example GET
     */
    get method() {
        return this.request.method;
    }
    /**
     * Get the base url from request.
     *
     * @example /users/1
     */
    get baseUrl() {
        return this.request.url.split('?')[0];
    }
    /**
     * Get the base url with host and port info from request.
     *
     * @example http://localhost:3030/users/1
     */
    get baseHostUrl() {
        return this.getHostUrlFor(this.baseUrl);
    }
    /**
     * Get the route url from request.
     *
     * @example /users/:id
     */
    get routeUrl() {
        return this.request.routerPath;
    }
    /**
     * Get the route url with host and port info from request.
     *
     * @example http://localhost:3030/users/:id
     */
    get routeHostUrl() {
        return this.getHostUrlFor(this.routeUrl);
    }
    /**
     * Get the original url from request.
     *
     * @example /users/1?query=true
     */
    get originalUrl() {
        return this.request.url;
    }
    /**
     * Get the original url with host and port info from request.
     *
     * @example /users/1?query=true
     */
    get originalHostUrl() {
        return this.getHostUrlFor(this.originalUrl);
    }
    /**
     * Get all body from request.
     */
    get body() {
        return this.request.body || {};
    }
    /**
     * Get all params from request.
     */
    get params() {
        return this.request.params || {};
    }
    /**
     * Get all queries from request.
     */
    get queries() {
        return this.request.query || {};
    }
    /**
     * Get all headers from request.
     */
    get headers() {
        return this.request.headers || {};
    }
    /**
     * Get the server port.
     */
    get port() {
        return this.getAddressInfo().port;
    }
    /**
     * Get the http version.
     */
    get version() {
        return this.request.raw.httpVersion;
    }
    /**
     * Get a value from the request params or the default value.
     */
    param(param, defaultValue) {
        return this.params[param] || defaultValue;
    }
    /**
     * Get a value from the request query param or the default value.
     */
    query(query, defaultValue) {
        return this.queries[query] || defaultValue;
    }
    /**
     * Get a value from the request header or the default value.
     */
    header(header, defaultValue) {
        return this.headers[header] || defaultValue;
    }
    /**
     * Get only the selected values from the request body.
     */
    only(keys) {
        const body = {};
        Object.keys(this.body).forEach(key => {
            if (!keys.includes(key)) {
                return;
            }
            body[key] = this.body[key];
        });
        return body;
    }
    /**
     * Get all the values from the request body except the selected ones.
     */
    except(keys) {
        const body = {};
        Object.keys(this.body).forEach(key => {
            if (keys.includes(key)) {
                return;
            }
            body[key] = this.body[key];
        });
        return body;
    }
    /**
     * Get a value from the request body or the default value.
     */
    input(key, defaultValue) {
        return this.payload(key, defaultValue);
    }
    /**
     * Get a value from the request body or the default value.
     */
    payload(key, defaultValue) {
        return Json.get(this.body, key, defaultValue);
    }
    /**
     * Add the hostname and port to the url.
     */
    getHostUrlFor(url) {
        let { address, port } = this.getAddressInfo();
        if (address === '::1') {
            address = '127.0.0.1';
        }
        if (!Is.Ip(address) && address !== 'localhost') {
            return `${this.protocol}://${address}${url}`;
        }
        return `${this.protocol}://${address}:${port}${url}`;
    }
    /**
     * Get the address info of the server. This method will return the
     * port used to listen the server, the family (IPv4, IPv6) and the
     * server address (127.0.0.1).
     */
    getAddressInfo() {
        return this.request.server.server.address();
    }
}
