// @adonisjs/validator/build/src/Validations/string/escape.js
"use strict";
/*
 * @adonisjs/validator
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.escape = void 0;
const escape_1 = __importDefault(require("validator/lib/escape"));
const helpers_1 = require("../../Validator/helpers");
const RULE_NAME = 'escape';
/**
 * Escape string value
 */
exports.escape = {
    compile: (0, helpers_1.wrapCompile)(RULE_NAME, [], () => {
        return {
            name: 'escape',
            async: false,
            allowUndefineds: false,
            compiledOptions: undefined,
        };
    }),
    validate(value, _, { mutate }) {
        if (typeof value !== 'string') {
            return;
        }
        mutate((0, escape_1.default)(value));
    },
};
