// @athenna/common/src/Helpers/Collection.js
/**
 * @athenna/common
 *
 * (c) João Lenon <lenon@athenna.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { Collection as CollectJS } from 'collect.js';
export class Collection extends CollectJS {
    /**
     * An alias for macro instance method:
     *
     * @example
     *  Collection.macro('upperAndTrim', (value) => {
     *    return value.trim().toUpperCase()
     *  })
     */
    static macro(name, fn) {
        return new Collection().macro(name, fn);
    }
    /**
     * Remove all duplicated values from the array.
     */
    removeDuplicated() {
        return [...new Set(this.all())];
    }
    /**
     * Execute the toResource method inside objects if exists.
     */
    toResource(criterias = {}) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        return this.all().map(item => item.toResource(criterias));
    }
}
Collection.prototype.order = Collection.prototype.sort;
Collection.prototype.orderBy = Collection.prototype.sortBy;
Collection.prototype.orderByDesc = Collection.prototype.sortByDesc;
Collection.prototype.orderDesc = Collection.prototype.sortDesc;
Collection.prototype.orderKeys = Collection.prototype.sortKeys;
Collection.prototype.orderKeysDesc = Collection.prototype.sortKeysDesc;
